added part2 topics (#2)
Programming-Club-IITM/GitPractice: README.md

12) Now merge the other branch with main branch using git merge myFirstBranch <br/>
13) Now push all the changes onto your forked github repository using git push origin/main <br/>
 
 Open your forked github repository and check whether all the changes have been updated. <br/>

## Log

`git log` is used to display the commit history of the repository. Try using this command and you will be able to see the commit history of the repository.

```bash
git log
```

However, the output of the above command may be too verbose. You can use the following command to see a more concise output.

```bash
git log --all --decorate --oneline --graph
```

In the current state, you should see a commit history that looks like this:

> ![Commit Log](/content/images/log_mid.png)

Reviewer: The commit messages are a bit misleading, could you change it such that they reflect which commit comes first? For example the last one maybe named "My Latest Commit" etc.

Author: Is it about `commit1` and `commit2` which are present in all the images? I guess I can simply not show that as it is not relevant

Reviewer 2: yes do that

    Notice that the commit history is now linear, and the commits from `mySecondBranch` have been reapplied on top of the `main` branch. However, the commit hashes will be different before and after the rebase.

### Rebase vs Merge

In cases where we want to combine the changes from one branch to another, we can use either `rebase` or `merge`.

| Merge                                                            | Rebase                                                          |
| ---------------------------------------------------------------- | -------------------------------------------------------------- |
| Creates a new commit that combines the changes from two branches | Reapplies the commits from one branch on top of another branch |
| Creates a non-linear commit history                              | Creates a linear commit history                                |
| Preserves the commit history of both branches                    | Rewrites the commit history of the branch being rebased |

Now that you have rebased the `mySecondBranch` branch on top of the `main` branch, use `git log` to compare the commit history with the one where you merged `myFirstBranch` with `main`.

> 

Reviewer: Better to have picture here as well. Do you use VSCode? there is a remarkable feature that shows you the commit graph. Something like this:
![image](https://github.com/Programming-Club-IITM/GitPractice/assets/73378019/4c5afa84-90b0-4c4b-829c-edf603d833be)
Including that here would help. Or including a general picture from the internet would also be fine, one that shows the difference between rebasing and merging clearly

Author: yes i do have vscode, but do not know about such a feature

I came accross `gitk` as well, but I am not sure whether all environments have it. Nevertheless, it isn't very pretty.

In this case however, the merge of `myFirstBranch` would be a fast-forward merge. So I couldn't get anything to visualize it like a three-way-merge.

Reviewer 2: I think it comes with gitlens/gitkraken idk which one exactly. Try to get it, super helpful. Put whatever pic seems fine. But include one. I would suggest getting this and then trying it out

Author: wow, its been 4 minutes since i installed gitlens and it is already quite impressive...
and yeah, i will add the necessary image

    ![Commit Log 3](/content/images/log_cherrypick_1.png)

## Interactive Rebase

`git rebase -i` is used to interactively rebase commits. This is useful when you want to modify, delete, or combine commits.

Consider the following scenario:

1) Switch to `mySecondBranch`. Add a new text file called `new5.txt` and add some text to it. Add and commit these changes.
2) Edit the `new2.txt` file and add some more text to it. Add and commit these changes.
3) Now, you should have the following commit history:

    ![Commit Log 4](/content/images/log_irebase.png)

4) You want to rebase `mySecondBranch` onto `main`, but you have already cherry-picked the commit that added `new3.txt` to the `main` branch.

    Now, you want the `new2.txt` file in your main branch, but it has two commits associated with it. You want to combine these two commits into a single commit. Moreover, now you do not want the `new5.txt` file anymore.

    Checkout to `mySecondBranch` and use the following command to interactively rebase the commits:

    ```bash

>     git rebase -i main

Reviewer: Include this statement as well ```git rebase -i HEAD~3``` before the rebasing onto main example. You don't have to write all the steps as they are all the same. Just mention a line or two about what this does, and when it is used

Author: ok, will do

    ```

>     You will see a list of commits that are being rebased. You can modify this list to combine, delete, or edit commits.

Reviewer: ```suggestion
    A new file will open and you will see that it has a list of commits that are being rebased. You can modify this list to combine, delete, or edit commits.
```

    In your case, you might see something like this:

    ![Interactive 1](/content/images/interactive1.png)

> 5) You need to edit this file to combine the two commits that added `new2.txt`. To do this, delete the line corresponding to the commit that `added new5.txt` and change the word `pick` to `squash` for the commit that `modified new2.txt`.

Reviewer: ```suggestion
5) You need to edit this file to combine the two commits that added `new2.txt`. To do this, change the word `pick` to `drop` corresponding to the commit that `added new5.txt` and change the word `pick` to `squash` for the commit that `modified new2.txt`.
```
A clearer way of doing the same thing. Also, separate the two things drop and squash into two lines for clarity. Also include a line that says that the squash commit will not exist and the changes will be reflected in the commit right before it.

If we merge `branch1` with `main` using the following commands:

```bash
git checkout main
git merge branch1
```

The commit history will look like this:

![Merge State](/content/images/merge_state.png)

Notice that the commit history is non-linear, and the changes from `branch1` have been combined with the `main` branch by creating a **new** commit. This is a three-way merge.

Instead, if we had rebased `branch1` onto `main` using the following commands:

```bash
git checkout branch1
git rebase main
```

The commit history would have looked like this:

![Rebase State](/content/images/rebase_state.png)

Notice that the commit history is linear, and the changes from `branch1` have been reapplied on top of the `main` branch. In order to get `main` up to date with the changes in `branch1`, we can now merge `branch1` with `main`.

```bash
git checkout main
git merge branch1
```

**Remember:** Now, it would be a fast-forward merge and the linear commit history would be maintained.

![Rebase->Merge State](/content/images/rebase->merge_state.png)

## Cherry-pick

> `git cherry-pick` is used to apply the changes introduced by some existing commits. It is useful when you want to apply a single commit from one branch to another.

Reviewer: ```suggestion
`git cherry-pick` is used to apply the changes introduced by some existing commits. It is useful when you want to apply a one or more commits from one branch to another.
```

    ```

3) Verify that the changes have been applied to the `main` branch by using the `git log` command. You can also verify this by checking that you now have the `new3.txt` file in the `main` branch, but not the `new2.txt` file.

    ![Commit Log 3](/content/images/log_cherrypick_1.png)

## Interactive Rebase

`git rebase -i` is used to interactively rebase commits. This is useful when you want to modify, delete, or combine commits.

Consider the following scenario:

1) Switch to `mySecondBranch`. Add a new text file called `new5.txt` and add some text to it. Add and commit these changes.
2) Edit the `new2.txt` file and add some more text to it. Add and commit these changes.
3) Now, you should have the following commit history:

    ![Commit Log 4](/content/images/log_irebase.png)

4) You want to rebase `mySecondBranch` onto `main`, but you have already cherry-picked the commit that added `new3.txt` to the `main` branch.

    Now, you want the `new2.txt` file in your main branch, but it has two commits associated with it. You want to combine these two commits into a single commit. Moreover, you do not want the `new5.txt` file anymore.

    Checkout to `mySecondBranch`:

>     At this point, if you wanted to modify the commit history of `mySecondBranch`, you can use: `git rebase -i HEAD~4`. Notice that `HEAD` refers to the current commit (the latest commit on `mySecondBranch`), and `HEAD~4` refers to the commit which is four commits **before** the current commit (in this case, the first commit on `mySecondBranch`). This command will not rebase onto `main`, but rather onto the commit denoted by `HEAD~4`. In this case, it will rebase onto itself, effectively allowing you to modify the commit history of `mySecondBranch` by interacting with the 4 commits on `mySecondBranch`.

Reviewer: ```suggestion
    At this point, if you wanted to modify the commit history of `mySecondBranch`, you can use: `git rebase -i HEAD~4`. Notice that `HEAD` refers to the current commit (the latest commit on `mySecondBranch`), and `HEAD~4` refers to the commit which is four commits **before** the current commit (in this case, the first commit on `mySecondBranch`). This command will not rebase onto `main`, but rather onto the commit denoted by `HEAD~4`. In this case, it will rebase onto itself, effectively allowing you to modify the commit history of `mySecondBranch` by interacting with the last 4 commits on `mySecondBranch`. 
Note that you need to be careful to not rebase with a number greater than 4 if you have committed for a maximum of 4 times in the current PR, else the commits in `main` before this PR will also be pulled in and will clutter the repository
```
